webhooks/virtualmachinegrouppublishrequest/mutation/virtualmachinegrouppublishrequest_mutator.go

```
@@ -0,0 +1,152 @@
+	}
+}
+
+type mutator struct {
+	client    ctrlclient.Client
+	converter runtime.UnstructuredConverter
+}
+
+func (m mutator) For() schema.GroupVersionKind {
+	return vmopv1.GroupVersion.WithKind(reflect.TypeOf(vmopv1.VirtualMachineGroupPublishRequest{}).Name())
+}
+
+func (m mutator) Mutate(ctx *pkgctx.WebhookRequestContext) admission.Response {
+	// this check is unnecessary but serves as a saft guard
+	if ctx.Op != admissionv1.Create {
+		return admission.Allowed("")
+	}
+
+	return m.mutateCrate(ctx)
+}
+
+// mutateCrate sets spec.source to the VM group publish request's name if it is omitted,
+// spec.target to a default content library if one exists, and spec.virtualMachines to
+// all the VMs under source VM group.
+func (m mutator) mutateCrate(ctx *pkgctx.WebhookRequestContext) admission.Response {
+	groupPubReq, err := m.vmGroupPublishRequestFromUnstructured(ctx.Obj)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if groupPubReq.Spec.Source == "" {
+		groupPubReq.Spec.Source = groupPubReq.Name
+	}
+
+	if groupPubReq.Spec.Target == "" {
+		defaultCL, err := common.RetrieveDefaultImagePublishContentLibrary(ctx, m.client, groupPubReq.Namespace)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return admission.Errored(http.StatusNotFound, err)
+			}
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		groupPubReq.Spec.Target = defaultCL.Name
+	}
+
+	if len(groupPubReq.Spec.VirtualMachines) == 0 {
+		vms, err := m.retrieveVMGroupMembers(ctx, ctrlclient.ObjectKey{Namespace: groupPubReq.Namespace, Name: groupPubReq.Spec.Source})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return admission.Errored(http.StatusNotFound, err)
+			}
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		groupPubReq.Spec.VirtualMachines = vms
+	}
+	// ensure vms are unique and sorted
+	groupPubReq.Spec.VirtualMachines = sets.List(sets.New(groupPubReq.Spec.VirtualMachines...))
```

I'm not sure if we want to change the whatever order this was created with. Like a user made a copy-pasta error that is basically swallowed here

The unique part is better asserted in the validation webhook, and should fail for dupes

---

yeah, was thinking that the API could be a little more tolerance but it makes sense to keep user input without modification.